### PR TITLE
Migrate Edit Navigation screen "Delete menu" button from `confirm()` …

### PR DESCRIPTION
**Pull request by chad1008**
Related: https://github.com/WordPress/gutenberg/pull/34153

## Description
This PR aims to migrate the beta Navigation Editor's `Delete menu` button away from the current `confirm()` implementation and instead use the new experimental `ConfirmDialog` component.

## How has this been tested?
Running WordPress 5.8.2 via `wp-env`:
1. Ensure the beta Navigation editor is activated from `Gutenberg > Experiments`
2. Open `Gutenberg > Navigation (beta)`
3. Create multiple menus (two or more)
4. Click the `Delete menu` button. Confirm there are no unexpected console errors and the new `ConfirmDialog` component is rendered instead of the browser's `confirm()` prompt.
5. Click the `Cancel` button. Confirm that the menu is still intact.
6. Click the `Delete menu` button again, and this time select `OK`
7. Confirm the menu in question is, in fact, deleted
8. Deleting one menu should automatically load the next menu that's in the database. Use `Delete menu` on all of them
9. Confirm that deleting the final/only menu returns you to the `Create your first menu` UI

Note, if you already have a menu saved from a previous session, attempting to delete it will result in an error. This doesn't appear to be related to this PR, and has been reported in #37794

I've tested in the latest Chrome, Firefox, and Safari.

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [ ] My code follows the accessibility standards. 
- [ ] I've tested my changes with keyboard and screen readers. 
- [ ] My code has proper inline documentation. 
- [ ] I've included developer documentation if appropriate. 
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal).

### CODE REVIEW

**Comment by chad1008** [comment]
Marking as `Draft` - there's a bug when cancel is clicked on the `ConfirmDialog` that prevents future clicks of the Delete menu button from being registered. I'll investigate and reopen once that's fixed!

**Comment by chad1008** [comment]
Fixed - state is now handled more effectively, and the confirmation dialog is reset after the cancel button is clicked.

**Comment by chad1008** [comment]
Updated this PR to simplify the implementation of ConfirmDialog (cc @ciampo)

File: packages/edit-navigation/src/components/sidebar/delete-menu.js
Comment (ciampo) [inline packages/edit-navigation/src/components/sidebar/delete-menu.js:30]: Small optimisation: you could look into memoizing this callback so that the function doesn't get recreated every time the `DeleteMenu` component renders (see [related comment](https://github.com/WordPress/gutenberg/pull/37491#discussion_r780804949))

File: packages/edit-navigation/src/components/sidebar/delete-menu.js
Comment (ciampo) [inline packages/edit-navigation/src/components/sidebar/delete-menu.js:37]: Similar comment re. memoizing this callback instead of re-creating for every render.

File: packages/edit-navigation/src/components/sidebar/delete-menu.js
Comment (ciampo) [inline packages/edit-navigation/src/components/sidebar/delete-menu.js]: Should we close the dialog when confirming it? The current implementation (using the native `confirm()`) seems to behave that way. (see also [related comment](https://github.com/WordPress/gutenberg/pull/37491#discussion_r780805958))

**Review by ciampo — COMMENTED** [review]
Hey @chad1008 , thank you for working on this!

My feedback is basically the same as [the one I left in PR 37491](https://github.com/WordPress/gutenberg/pull/37491#pullrequestreview-847280338), including the consideration about unit/e2e tests.

File: packages/edit-navigation/src/components/sidebar/delete-menu.js
Comment (kevin940726) [reply packages/edit-navigation/src/components/sidebar/delete-menu.js:30]: It isn't necessarily a performance optimization though, [sometimes it might get worse](https://kentcdodds.com/blog/usememo-and-usecallback). I'd prefer to keep it simple until we audit it.

Comment (ciampo) [reply packages/edit-navigation/src/components/sidebar/delete-menu.js:30]: Sure, we can keep it as-is unless an optimisation is needed

Comment (chad1008) [reply packages/edit-navigation/src/components/sidebar/delete-menu.js:30]: @ciampo I know several of your other comments on this PR (and the other's you've reviewed for me) also touched on memoizing callbacks. They're all similar callbacks, does the approach of holding off until optimization is needed apply to those recommendations as well?

Comment (ciampo) [reply packages/edit-navigation/src/components/sidebar/delete-menu.js:30]: Let's hold off for now, it's always something that we can add later on — it will also allow you to focus on other aspects of the feedback first :)

**Comment by ciampo** [comment]
Update:

- I marked all of the memoization-related conversations as resolved for now, as discussed in [this comment](https://github.com/WordPress/gutenberg/pull/37492#discussion_r780807376)
- there's still an [open comment](https://github.com/WordPress/gutenberg/pull/37492#discussion_r780808161) that needs addressing
- not sure if you managed to look into fixing any broken e2e/unit tests (if any) and potentially adding any new ones

Even with the last comment addressed, I'd still be reluctant to merge this PR until we find a good fix for the `z-index` issue (see https://github.com/WordPress/gutenberg/pull/37959).

File: packages/edit-navigation/src/components/sidebar/delete-menu.js
Comment (chad1008) [reply packages/edit-navigation/src/components/sidebar/delete-menu.js]: Done in 4667075!

Comment (chad1008) [reply packages/edit-navigation/src/components/sidebar/delete-menu.js]: done in 4667075!

File: packages/edit-navigation/src/components/sidebar/delete-menu.js
Comment (kevin940726) [inline packages/edit-navigation/src/components/sidebar/delete-menu.js:17]: Maybe we should `await` for `onDeleteMenu()` to finish? (I think it returns a promise?)

Bonus point if we can also catch errors here and display the message somewhere, but that's probably best for a different PR.

Another micro-optimization we can make is to show a loading state when the menu is still being deleted. Something like disabling the confirm button and showing a loading icon on the button. But again, probably best for a follow-up PR (I don't think it's possible now with the current API in `__experimentalConfirmDialog` anyway).

These are all just nitpicking though, I think we can merge this and work on those if needed.

**Review by kevin940726 — APPROVED** [review]

File: packages/edit-navigation/src/components/sidebar/delete-menu.js
Comment (ciampo) [reply packages/edit-navigation/src/components/sidebar/delete-menu.js:17]: > Maybe we should `await` for `onDeleteMenu()` to finish? (I think it returns a promise?)`

Hard to say for sure without any docs in this component.

I did a quick search in the codebase, and it looks like the only usage is in [`packages/edit-navigation/src/components/layout/index.js`](https://github.com/WordPress/gutenberg/blob/11294e370b29a237080b946e405c61ed75fa9509/packages/edit-navigation/src/components/layout/index.js#L195). In that case, the `deleteMenu` function comes from the [`useNavigationEditor()` hook](https://github.com/WordPress/gutenberg/blob/ca4cadf63e28998e463615eab269b5b56cc26bd8/packages/edit-navigation/src/hooks/use-navigation-editor.js#L77-L97), and it is indeed an `async` function.

So yeah, we could technically

```js
await onDeleteMenu();
setShowConfirmDialog( false );
```

> Bonus point if we can also catch errors here and display the message somewhere, but that's probably best for a different PR.

Given the [delete function definition](https://github.com/WordPress/gutenberg/blob/ca4cadf63e28998e463615eab269b5b56cc26bd8/packages/edit-navigation/src/hooks/use-navigation-editor.js#L77-L97), it looks like errors are handled via the `createErrorNotice` function, which seems to be connected to the notices store — I'm not familiar with what's the best "gutenberg" way to handle that?

> Another micro-optimization we can make is to show a loading state when the menu is still being deleted. Something like disabling the confirm button and showing a loading icon on the button. But again, probably best for a follow-up PR (I don't think it's possible now with the current API in __experimentalConfirmDialog anyway).

We should have access to the loading state via the `isMenuBeingDeleted` prop passed to this component, but I don't think that the `ConfirmDialog` supports a way to disable its confirm button from outside the component (@fullofcaffeine can you confirm?)

Comment (kevin940726) [reply packages/edit-navigation/src/components/sidebar/delete-menu.js:17]: c.c. @adamziel as this is a good example of lacking a clear and simple way to do loading state + error handling with the current API 😆 . (#38135)

Comment (fullofcaffeine) [reply packages/edit-navigation/src/components/sidebar/delete-menu.js:17]: > supports a way to disable its confirm button from outside the component (@fullofcaffeine can you confirm?)

You're right, that's not supported at the moment. I think it's a great improvement idea! What would that API look like do you think? A simple `isLoading` boolean prop that would disable the button and show a loading spinner when `true`?

Comment (kevin940726) [reply packages/edit-navigation/src/components/sidebar/delete-menu.js:17]: > A simple `isLoading` boolean prop that would disable the button and show a loading spinner when `true`?

Yeah, I think so. Since it should be a common feature, we can allow a simple prop like `isLoading` or `disabled` to do the job.

Comment (ciampo) [reply packages/edit-navigation/src/components/sidebar/delete-menu.js:17]: >  A simple `isLoading` boolean prop that would disable the button and show a loading spinner when true?

I think I like the sound of `isDisabled` more as it is closer to what we want to achieve — disable interactions with the modal. Let's also add a `is` in front of the prop to make it clear that it's a boolean flag.

We also have to disable pressing the ESC key and clicking on the background, and we'll have to make this feature work both in controlled and uncontrolled mode. We can discuss further these details and refine the UX separately in the PR implementing the feature.

**Comment by ciampo** [comment]
Since this PR has been merged by @draganescu, let's add any remaining changes (potential e2e test improvements, and the points discussed in [this conversation](https://github.com/WordPress/gutenberg/pull/37492#discussion_r791483132)) in separate follow-up PRs (cc @chad1008 ).

**Pull request merged**
